src/solc_0.8/defi/rules/ContributionRules.sol

```diff
@@ -0,0 +1,194 @@
+//SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.2;
+
+import {Ownable} from "@openzeppelin/contracts-0.8/access/Ownable.sol";
+import {Address} from "@openzeppelin/contracts-0.8/utils/Address.sol";
+import {SafeMathWithRequire} from "../../common/Libraries/SafeMathWithRequire.sol";
+import {IERC721} from "@openzeppelin/contracts-0.8/token/ERC721/IERC721.sol";
+import {IERC1155} from "@openzeppelin/contracts-0.8/token/ERC1155/IERC1155.sol";
+
+contract ContributionRules is Ownable {
+    using Address for address;
+    uint256 internal constant DECIMALS_7 = 10000000;
+    uint256 internal constant MIDPOINT_9 = 500000000;
+    uint256 internal constant NFT_FACTOR_6 = 10000;
+    uint256 internal constant NFT_CONSTANT_3 = 9000;
+    uint256 internal constant ROOT3_FACTOR = 697;
+
+    struct MultiplierERC721 {
+        uint256[] ids;
+        uint256[] multiplier;
+        uint256 index;
+        bool balanceOf;
+    }
+
+    struct MultiplierERC1155 {
+        IERC1155 assetContract;
+        uint256[] ids;
+        uint256[] multiplier;
+        uint256 index;
+    }
+
+    mapping(IERC721 => MultiplierERC721) private _listERC721;
+    mapping(IERC1155 => MultiplierERC1155) private _listERC1155;
+    IERC721[] internal _listERC721Index;
+    IERC1155[] internal _listERC1155Index;
+
+    IERC721 public landMultiplierContract;
+
+    constructor(IERC721 _landMultiplierContract) {
```

Comment:
no land

Reply:
Also maybe we could pass all the MultiplierERC721 & MultiplierERC1155 in the contructor ?

Reply:
done -> https://github.com/thesandboxgame/sandbox-smart-contracts/pull/605/commits/8b5415cc6935ba36ebc83b6fedd4816f99c7b8da

Reply:
For the constructor, I think we should set this after the deployment. Or we gonna add some more few loops to the contract to set the multipliers.